internal/kafka/internal/workers/cluster_mgrs/dynamic_scaleup_mgr.go

```diff
@@ -124,3 +144,147 @@ func (m *DynamicScaleUpManager) reconcileClustersForRegions() []error {
+//        * Clusters that are still not ready to accept kafka instance but that
+//          should eventually accept them (like accepted state for example)
+//          are included
+// Otherwise false is returned.
+// Note: Clusters in the 'failed' state are excluded from the calculations.
+func (m *DynamicScaleUpManager) shouldScaleUpNewClusterForInstanceType(
+	provider string, region string,
+	instanceTypeName string,
+	instanceTypeConfig *config.InstanceTypeConfig,
+	kafkaStreamingUnitCountPerClusterList services.KafkaStreamingUnitCountPerClusterList) (bool, error) {
+
+	kafkaInstanceTypeConfig, err := m.KafkaConfig.SupportedInstanceTypes.Configuration.GetKafkaInstanceTypeByID(instanceTypeName)
+	if err != nil {
+		return false, err
+	}
+	biggestKafkaInstanceSizeCapacityConsumption := -1
+	maxKafkaInstanceSizeConfig := kafkaInstanceTypeConfig.GetBiggestCapacityConsumedSize()
+	if maxKafkaInstanceSizeConfig != nil {
+		biggestKafkaInstanceSizeCapacityConsumption = maxKafkaInstanceSizeConfig.CapacityConsumed
+	}
+
+	consumedStreamingUnitsInRegion := 0
+	maxStreamingUnitsInRegion := 0
+	atLeastOneClusterHasCapacityForLargestInstanceType := false
+
+	scaleUpActionIsOngoing := false
+	clusterStatesTowardReadyState := []string{
+		api.ClusterProvisioning.String(), api.ClusterProvisioned.String(),
+		api.ClusterAccepted.String(), api.ClusterWaitingForKasFleetShardOperator.String(),
+	}
+
+	for _, kafkaStreamingUnitCountPerCluster := range kafkaStreamingUnitCountPerClusterList {
+		if kafkaStreamingUnitCountPerCluster.CloudProvider == provider &&
+			kafkaStreamingUnitCountPerCluster.Region == region &&
+			kafkaStreamingUnitCountPerCluster.InstanceType == instanceTypeName {
+
+			if arrays.Contains(clusterStatesTowardReadyState, kafkaStreamingUnitCountPerCluster.Status) {
+				scaleUpActionIsOngoing = true
+			}
+
+			if kafkaStreamingUnitCountPerCluster.FreeStreamingUnits() >= int32(biggestKafkaInstanceSizeCapacityConsumption) {
+				atLeastOneClusterHasCapacityForLargestInstanceType = true
+			}
+
+			consumedStreamingUnitsInRegion = consumedStreamingUnitsInRegion + int(kafkaStreamingUnitCountPerCluster.Count)
+			if kafkaStreamingUnitCountPerCluster.Status != api.ClusterDeprovisioning.String() &&
+				kafkaStreamingUnitCountPerCluster.Status != api.ClusterCleanup.String() {
+				maxStreamingUnitsInRegion = maxStreamingUnitsInRegion + int(kafkaStreamingUnitCountPerCluster.MaxUnits)
+			}
+		}
+	}
+
+	freeStreamingUnitsInRegion := maxStreamingUnitsInRegion - consumedStreamingUnitsInRegion
+	capacitySlackInRegion := instanceTypeConfig.MinAvailableCapacitySlackStreamingUnits
+	notEnoughCapacitySlackInRegion := false
+	if freeStreamingUnitsInRegion < capacitySlackInRegion { // TODO is it <= or < ?
+		notEnoughCapacitySlackInRegion = true
+	}
+
+	streamingUnitsLimitInRegion := instanceTypeConfig.Limit
+	regionLimitReached := false
+	if streamingUnitsLimitInRegion != nil && consumedStreamingUnitsInRegion >= *streamingUnitsLimitInRegion {
```

Comment:
if region limits is `nil`, 

do we want to compare `freeStreamingUnitsInRegion > 0` and used that to set `regionLimitReached` to `true`?

That could be redundant with`atLeastOneClusterHasCapacityForLargestInstanceType` but it really ensures that we've not consumed all current available max limits in the region for cases where we do not have region limit set.

Reply:
If region limits is nil then regionLimitReached should never be true

Reply:
Makes sense. 

> do we want to compare freeStreamingUnitsInRegion > 0 and used that to set regionLimitReached to true?

> That could be redundant withatLeastOneClusterHasCapacityForLargestInstanceType but it really ensures that we've not consumed all current available max limits in the region for cases where we do not have region limit set.


what I was saying above is 100% percent covered by https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/2b7ba102694ac844d1d5421fc02636c43f38a554/internal/kafka/internal/workers/cluster_mgrs/dynamic_scaleup_mgr.go#L234

Reply:
:+1: . Marking this as resolved then.